feature/work-schedule-BJS2-96030

Deathroels opened the pull request.

Deathroels committed:
BJS2-96030: Рабочий график
-Добавил и записал Dto для графика

Deathroels committed:
BJS2-96030: Рабочий график
-Добавил mapper для WorkSchedule и WorkScheduleDto

Deathroels committed:
BJS2-96030: Рабочий график
-Создал интерфейс WorkScheduleService c методом addWorkSchedule

Deathroels committed:
BJS2-96030: Рабочий график
-исправил имена переменных в Dto классах

Deathroels committed:
BJS2-96030: Рабочий график
-добавил реализацию addWorkSchedule у WorkScheduleServiceImpl

Deathroels committed:
BJS2-96030: Рабочий график
-добавил класс валидатор бизнес-логики, использовал в сервисе
-добавил аннотации для валидации в дто
-добавил контроллер с методом добавления рабочего графика пользователю

Deathroels committed:
BJS2-96030: Рабочий график
-добавил в контроллер метод для обновления рабочего графика и получение графика по его id;

Deathroels committed:
BJS2-96030: Рабочий график
-изменил логику обновления графика, с учетом опечаток в задаче

Deathroels committed:
BJS2-96030: Рабочий график
-добавил логи в сервис
-добавил аннотацию бина, для валидатора

Deathroels committed:
BJS2-96030: Рабочий график
-фикс логикик проверки на изменения в чужом графике

Deathroels committed:
BJS2-96030:
-set show-sql to false

Deathroels committed:
BJS2-96030:
-implement REST endpoints for one WorkSchedule method

Deathroels committed:
BJS2-96030: Рабочий график
-перенес валидаторы в другой пакет

Deathroels committed:
BJS2-96030: Рабочий график
-убрал ненужную аннотацию в валидатотре (заменил на статический метод, т.к. нету полей с зависимостями и состояниями)

Deathroels committed:
BJS2-96030: Рабочий график
-убрал ненужную валилацию в дто(проверка на уровне сущности, проверка имела бы смысл при приеме данных)

Deathroels committed:
BJS2-96030: Рабочий график
-убрал неиспользуемый интерфейс сервиса, оставил одну реализацию
-валидация в методах изменена на исп. статического метода

Deathroels committed:
BJS2-96030: Рабочий график
-убрал отдельный мапер на обновление и расширил текущий
-изменил логику обновления в сервисе, с учетом изменений мапера и неправильной проверки в if блоке
-изменил if блок воизбежании проблем с сравнением объектотв

Deathroels committed:
BJS2-96030: Рабочий график
-изменил вывод валидации на английский язык

Deathroels committed:
BJS2-96030: Рабочий график
-добавил оставшиеся REST endpoints, для контроллера

Deathroels committed:
BJS2-96030: Рабочий график
-добавил тесты связанные с работой сервиса

Deathroels committed:
BJS2-96030: Рабочий график
-fix checkstyle

Brocksvar commented on src/main/java/school/faang/user_service/controller/workschedule/WorkScheduleController.java:
```diff
@@ -0,0 +1,38 @@
+@RequiredArgsConstructor
+public class WorkScheduleController {
+    private final WorkScheduleService workScheduleService;
+    private final UserContext userContext;
```
UserContext лучше вызывать в сервисе

Brocksvar commented on src/main/java/school/faang/user_service/controller/workschedule/WorkScheduleController.java:
```diff
@@ -0,0 +1,38 @@
+        return workScheduleService.addWorkSchedule(userContext.getUserId(), workScheduleDto);
+    }
+
+    @PutMapping("/{id}")
```
Лучше использовать @PatchMapping, т.к. обновляется не вся сущность, а её часть

Brocksvar commented on src/main/java/school/faang/user_service/mapper/WorkScheduleMapper.java:
```diff
@@ -0,0 +1,16 @@
+import school.faang.user_service.dto.workschedule.WorkScheduleDto;
+import school.faang.user_service.entity.user.WorkSchedule;
+
+@Mapper(componentModel = "spring")
```
добавь параметр `nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE`, чтобы метод `updateWorkSchedule` не заменил существующие поля на `null`

Brocksvar commented on src/main/java/school/faang/user_service/service/validator/UpdateWorkScheduleValidator.java:
```diff
@@ -0,0 +1,18 @@
+import school.faang.user_service.dto.workschedule.UpdateWorkScheduleDto;
+import school.faang.user_service.exception.DataValidationException;
+
+public class UpdateWorkScheduleValidator {
```
я бы сделал один класс валидатор со всеми методами. Это не ошибка, но кажется что так удобнее.

Scherbi1 commented on src/main/java/school/faang/user_service/service/workschedule/WorkScheduleService.java:
```diff
@@ -0,0 +1,61 @@
+    private final WorkScheduleRepository workScheduleRepository;
+    private final WorkScheduleMapper workScheduleMapper;
+
+    public WorkScheduleDto addWorkSchedule(long userId, WorkScheduleDto workScheduleDto) {
```
Нужно сделать проверку на null. Если какие-то значения из workScheduleDto будут null, то  что произойдет в таком случае? 
Можешь либо тут проверить, но  я бы использовал джакарту (можешь написать в тг подробнее расскажу, как можно сделать)

Scherbi1 replied:
В сервисе лучше сразу работать с entity,  а маппить в котроллере или cделать прослойку между котроллером и сервисом, где будешь производить маппинг
Создай папку facade в папке controller и там уже класс, который будет маппить

Deathroels committed:
BJS2-96030:
-change put mapping for patch mapping

Deathroels committed:
BJS2-96030: Рабочий график
-перенес UserContext из контроллера в сервис

Deathroels committed:
BJS2-96030: Рабочий график
-исправлено поведение мапера (при обновлении, текущие поля в сущности, не меняются на поля переданного дто, если там значение null)

Deathroels committed:
BJS2-96030: Рабочий график
-добавлен общий расширеяемый валиадатотр для типов дто, с которыми идет взаимодействие

Deathroels committed:
BJS2-96030: Рабочий график
-добавил валидацию через Jakarta, для дто, которая создает график
-переместил маппинг и валидацию на отдельный уроведь - фасад

Deathroels committed:
BJS2-96030: Рабочий график
-добавил валидацию через Jakarta, для дто, которая создает график
-переместил маппинг и валидацию на отдельный уроведь - фасад

Deathroels committed:
BJS2-96030: Рабочий график
-Изменил тест валидатора

Deathroels committed:
BJS2-96030: Рабочий график
-Изменил тест валидатора

20cognosce commented on src/main/java/school/faang/user_service/controller/facade/WorkScheduleFacade.java:
```diff
@@ -0,0 +1,46 @@
+
+        WorkSchedule updatedWorkSchedule = workScheduleService.save(workSchedule);
+
+        return workScheduleMapper.toWorkScheduleDto(updatedWorkSchedule);
```
```
WorkSchedule workSchedule = workScheduleService.updateWorkSchedule(workScheduleId, updateWorkScheduleDto);
return workScheduleMapper.toWorkScheduleDto(updatedWorkSchedule);
```

Deathroels committed:
BJS2-96030: Рабочий график
-перенес валидацию jakarta в дто для создания графика, удалил из обычной дто
-убрал лишнюю имплементацию интерфейса валидатора

Deathroels committed:
BJS2-96030: Рабочий график
-вернул валидацию и бизнес логику в сервис, оставил в фасаде только маппинг
-переделал по необходимости валидатор
-поправил тест

Deathroels committed:
BJS2-96030: Рабочий график
-вернул валидацию и бизнес логику в сервис, оставил в фасаде только маппинг
-переделал по необходимости валидатор

20cognosce commented on src/main/java/school/faang/user_service/controller/workschedule/WorkScheduleController.java:
```diff
@@ -0,0 +1,37 @@
+@RequestMapping("/work-schedules")
+@RestController
+@RequiredArgsConstructor
+public class WorkScheduleController {
```
А удаление

20cognosce commented on src/main/java/school/faang/user_service/dto/workschedule/CreateWorkScheduleDto.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.time.LocalTime;
+
+public record CreateWorkScheduleDto(
```
WorkScheduleCreateDto
WorkScheduleUpdateDto

20cognosce commented on src/main/java/school/faang/user_service/dto/workschedule/WorkScheduleUpdateDto.java:
```diff
@@ -0,0 +1,23 @@
+        LocalTime endLunch,
+
+        @Nullable
+        String timezone) {
```
а какая по умолчанию тогда?

20cognosce commented on src/main/java/school/faang/user_service/dto/workschedule/UpdateWorkScheduleDto.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.time.LocalTime;
+
+public record UpdateWorkScheduleDto(
```
.

20cognosce commented on src/main/java/school/faang/user_service/dto/workschedule/UpdateWorkScheduleDto.java:
```diff
@@ -0,0 +1,11 @@
+package school.faang.user_service.dto.workschedule;
+
+import java.time.LocalTime;
+
+public record UpdateWorkScheduleDto(
+        LocalTime startTime,
+        LocalTime endTime,
+        LocalTime startLunch,
+        LocalTime endLunch,
+        String timezone) {
```
@Nullable

20cognosce commented on src/main/java/school/faang/user_service/service/validator/WorkScheduleValidator.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private static void validateTimeSequence(LocalTime t1, LocalTime t2, LocalTime t3, LocalTime t4) {
+        if (!(t1.isBefore(t2) && t2.isBefore(t3) && t3.isBefore(t4))) {
```
может лучше разбить на 3 проверки? и детализировать, какая именно из проверок стрельнула

20cognosce commented on src/main/java/school/faang/user_service/controller/facade/WorkScheduleFacade.java:
```diff
@@ -0,0 +1,40 @@
+    public WorkScheduleDto addWorkSchedule(CreateWorkScheduleDto createWorkScheduleDto) {
+        WorkSchedule workSchedule = workScheduleMapper.toCreateWorkSchedule(createWorkScheduleDto);
+        WorkSchedule savedWorkSchedule = workScheduleService.addWorkSchedule(workSchedule);
+        log.info("WorkSchedule for user added");
```
смысл этого лога, что мы узнали

20cognosce commented on src/main/java/school/faang/user_service/controller/facade/WorkScheduleFacade.java:
```diff
@@ -0,0 +1,40 @@
+
+    public WorkScheduleDto updateWorkSchedule(long workScheduleId, UpdateWorkScheduleDto updateWorkScheduleDto) {
+        WorkSchedule workSchedule = workScheduleService.updateWorkSchedule(workScheduleId, updateWorkScheduleDto);
+        log.info("WorkSchedule was updated");
```
что ты залогировал? ничего

20cognosce commented on src/main/java/school/faang/user_service/service/workschedule/WorkScheduleService.java:
```diff
@@ -0,0 +1,59 @@
+        workSchedule.setUser(user);
+
+        return workScheduleRepository.save(workSchedule);
+    }
```
гуд
но апдейт и создание лучше в сервисе делать, чтобы это было явно, не через MapStruct

20cognosce commented on src/main/java/school/faang/user_service/controller/facade/WorkScheduleFacade.java:
```diff
@@ -0,0 +1,40 @@
+    private final WorkScheduleMapper workScheduleMapper;
+
+    public WorkScheduleDto addWorkSchedule(CreateWorkScheduleDto createWorkScheduleDto) {
+        WorkSchedule workSchedule = workScheduleMapper.toCreateWorkSchedule(createWorkScheduleDto);
```
почему toCreateWorkSchedule?
ты же наоборот, из Create приводишь к сущности

и с update аналогично

20cognosce commented on src/main/java/school/faang/user_service/service/workschedule/WorkScheduleService.java:
```diff
@@ -0,0 +1,59 @@
+        WorkSchedule workSchedule = getById(workScheduleId);
+        long userId = userContext.getUserId();
+
+        if (!workSchedule.getUser().getId().equals(userId)) {
```
Objects.equals

20cognosce commented on src/main/java/school/faang/user_service/mapper/WorkScheduleMapper.java:
```diff
@@ -0,0 +1,16 @@
+
+    WorkSchedule toWorkSchedule(WorkScheduleDto workScheduleDto);
+
+    WorkScheduleDto toWorkScheduleDto(WorkSchedule workSchedule);
```
оставь только этот метод в мапере

20cognosce commented on src/main/java/school/faang/user_service/service/workschedule/WorkScheduleService.java:
```diff
@@ -0,0 +1,59 @@
+        entity.setEndTime(dto.endTime());
+        entity.setStartLunch(dto.startLunch());
+        entity.setEndLunch(dto.endLunch());
+        entity.setTimezone(dto.timezone());
```
ну а где проверки, что поле не null? я хочу обновить только время старта обеда, и как?

20cognosce commented on src/main/java/school/faang/user_service/service/workschedule/WorkScheduleService.java:
```diff
@@ -0,0 +1,59 @@
+    private final WorkScheduleRepository workScheduleRepository;
+    private final UserContext userContext;
+
+    public WorkSchedule addWorkSchedule(WorkSchedule workSchedule) {
```
ну так какая таймзона будет по умолчанию, если я отправлю null?

Deathroels replied:
Добавил в дто инициализацию через конструктор для timezone == null будет менятся на "UTC"

20cognosce reviewed (changes requested):
Доделываем

Deathroels committed:
BJS2-96035:
-убрал все метода из маппера кроме toWorkScheduleDto

Deathroels committed:
BJS2-96035:
-убрал log.info не имеющие смысла

Deathroels committed:
BJS2-96035:
-вместо equals, objects.equals

Deathroels committed:
BJS2-96035:
-вместо обобщенной валиадции сделал 3 разных, чтобы знать какая именно сработала

Deathroels committed:
BJS2-96035:
-переименовал dto
-добавил дефолтное значение для timezone
-WorkScheduleUpdateDto добавил аннотации jakarta

Deathroels committed:
BJS2-96035:
- добавил @Valid в контроллере, для валиадации дто для создания.
-изменены аннотации дто для создания
-убрали лишнюю работу в фасаде
-проверка на пустые поля при обновлении графика
-удалены неиспользуемые поля в валиадаторе

Deathroels committed:
BJS2-96035:
-добавлено удаление графика

Deathroels committed:
BJS2-96035:
-изменил логику обновления с соблюдением бизнес логики и обработкой null значений

Deathroels committed:
BJS2-96035:
-добавил @Builder на дто

Deathroels committed:
BJS2-96035:
-исправлен тест WorkScheduleValidatorTest

Deathroels committed:
BJS2-96035:
-каркас старого теста сервиса добавил, переделать

Deathroels committed:
BJS2-96030: delete broken test

Deathroels committed:
BJS2-96035: fixes

Deathroels committed:
BJS2-96035: add service tests

Deathroels committed:
BJS2-96035: fix dto logic

Deathroels committed:
Merge branch 'feature/work-schedule-tests-BJS2-96035' into feature/work-schedule-BJS2-96030

# Conflicts:
#	src/test/java/school/faang/user_service/service/validator/WorkScheduleValidatorTest.java

Deathroels committed:
BJS2-96035: fix checkstyle

Deathroels committed:
BJS2-96030: fix checkstyle

Deathroels committed:
Merge branch 'cerberus-master-stream12' into feature/work-schedule-BJS2-96030

Deathroels committed:
BJS2-96030: fix packages

20cognosce commented on src/main/java/school/faang/user_service/service/workschedule/validator/WorkScheduleValidator.java:
```diff
@@ -0,0 +1,49 @@
+
+public class WorkScheduleValidator {
+    public static void validateEntity(WorkSchedule workSchedule) {
+        validateTimeSequence(
+                workSchedule.getStartTime(),
+                workSchedule.getStartLunch(),
+                workSchedule.getEndLunch(),
+                workSchedule.getEndTime()
+        );
+    }
+
+    public static void validate(WorkScheduleDto dto) {
+        validateTimeSequence(
+                dto.startTime(),
+                dto.startLunch(),
+                dto.endLunch(),
+                dto.endTime()
+        );
+    }
+
+    public static void validateForCreate(WorkScheduleCreateDto dto) {
+        validateTimeSequence(
+                dto.startTime(),
+                dto.startLunch(),
+                dto.endLunch(),
+                dto.endTime()
+        );
+    }
```
дублирование кода, можно 1 метод сделать для валидации

20cognosce reviewed (approved):
Принято!
